anomdtct/anomdtct/data.py

```diff
@@ -0,0 +1,470 @@
+        ),
+        geo_t AS (
+          SELECT
+            client_id,
+            geo
+          FROM (
+            SELECT
+              client_id,
+              CONCAT(
+                country, ":",
+                geo_subdivision1, ":",
+                geo_subdivision2, ":",
+                cd_t.city
+              ) AS geo
+            FROM `moz-fx-data-shared-prod.telemetry.clients_first_seen` AS cd_t
+            INNER JOIN (SELECT geo FROM top_cities_t) AS top_cities_t
+            ON CONCAT(
+                country, ":",
+                geo_subdivision1, ":",
+                geo_subdivision2, ":",
+                city
+              ) = top_cities_t.geo
+          )
+        )
+        SELECT
+          submission_date AS date,
+          AVG(active_hours_sum) AS value,
+          COUNT(client_id) AS dau,
+          geo_t.geo AS geo
+        FROM (
+            SELECT
+              active_hours_sum,
+              client_id,
+              submission_date,
+            FROM `moz-fx-data-shared-prod.telemetry.clients_daily`
+            WHERE
+              submission_date = '{submission_date}'
+              AND attribution.content IS NOT NULL
+          )
+        INNER JOIN geo_t
+        USING(client_id)
+        GROUP BY geo, submission_date
+        ''',
+    "light_funnel_mean_active_hours_per_profile_country": '''
+        WITH geo_t AS (
+          SELECT
+            client_id,
+            geo
+          FROM (
+            SELECT
+              client_id,
+              country AS geo,
+            FROM `moz-fx-data-shared-prod.telemetry.clients_first_seen` AS cd_t
+          )
+        )
+        SELECT
+          submission_date AS date,
+          AVG(active_hours_sum) AS value,
+          COUNT(client_id) AS dau,
+          geo_t.geo AS geo
+        FROM (
+            SELECT
+              active_hours_sum,
+              client_id,
+              submission_date,
+            FROM `moz-fx-data-shared-prod.telemetry.clients_daily`
+            WHERE
+              submission_date = '{submission_date}'
+              AND attribution.content IS NOT NULL
+          )
+        INNER JOIN geo_t
+        USING(client_id)
+        GROUP BY geo, submission_date
+        '''
+}
+
+
+def get_raw_data(bq_client, bq_storage_client, metric):
+    return bq_client.query(
+        RAW_DATA_QUERIES[metric]
+    ).result().to_dataframe(bqstorage_client=bq_storage_client)
+
+
+def prepare_training_data(data, training_start, training_end):
+    clean_data = {}
+    clean_training_data = {}
+    # Suppress any geoXdate with less than 5000 profiles as per minimum
+    # aggregation standards for the policy this data will be released under.
+    data = data[data.dau >= 5000].drop(columns=["dau"])
+    for c in data.geo.unique():
+        # We don't want to include a region unless we have at least about
+        # two years of training data for the model
+        if (len(data.query("geo==@c")) < 600):
+            continue
```

Comment:
@jmccrosky What happens if the model tries to perform forecasting for a region that it hasn't seen any training data for? Is it just discarded? This might be relevant in the `prepare_data` function below.

Reply:
The forecast function takes a dictionary as its data argument and expects training data to exist for every key in that dictionary.  If there are insufficient rows in the training data, it will skip that key though.  Looks like you've changed things a lot though :)  I think you're probably handling it fine.